proof/refine/ARM/StateRelation.thy

```diff
@@ -385,11 +409,55 @@ where
   "sched_act_relation choose_new_thread a' = (a' = ChooseNewThread)" |
   "sched_act_relation (switch_thread x) a' = (a' = SwitchToThread x)"
 
-definition
-  ready_queues_relation :: "(Deterministic_A.domain \<Rightarrow> Structures_A.priority \<Rightarrow> Deterministic_A.ready_queue)
-                         \<Rightarrow> (domain \<times> priority \<Rightarrow> KernelStateData_H.ready_queue) \<Rightarrow> bool"
-where
-  "ready_queues_relation qs qs' \<equiv> \<forall>d p. (qs d p = qs' (d, p))"
+definition queue_end_valid :: "obj_ref list \<Rightarrow> tcb_queue \<Rightarrow> bool" where
+  "queue_end_valid ts q \<equiv>
+     (ts = [] \<longrightarrow> tcbQueueEnd q = None) \<and> (ts \<noteq> [] \<longrightarrow> tcbQueueEnd q = Some (last ts))"
+
+definition prev_queue_head :: "tcb_queue \<Rightarrow> (obj_ref \<rightharpoonup> 'a) \<Rightarrow> bool" where
+  "prev_queue_head q prevs \<equiv> \<forall>head. tcbQueueHead q = Some head \<longrightarrow> prevs head = None"
+
+lemma prev_queue_head_heap_upd:
+  "\<lbrakk>prev_queue_head q prevs; Some r \<noteq> tcbQueueHead q\<rbrakk> \<Longrightarrow> prev_queue_head q (prevs(r := x))"
+  by (clarsimp simp: prev_queue_head_def)
+
+definition list_queue_relation ::
+  "obj_ref list \<Rightarrow> tcb_queue \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref) \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref) \<Rightarrow> bool"
+  where
+  "list_queue_relation ts q nexts prevs \<equiv>
+     heap_ls nexts (tcbQueueHead q) ts \<and> queue_end_valid ts q \<and> prev_queue_head q prevs"
+
+lemma list_queue_relation_nil:
+  "list_queue_relation ts q nexts prevs \<Longrightarrow> ts = [] \<longleftrightarrow> tcbQueueEmpty q"
+  by (fastforce dest: heap_path_head simp: tcbQueueEmpty_def list_queue_relation_def)
+
+abbreviation ready_queue_relation ::
+  "Deterministic_A.domain \<Rightarrow> Structures_A.priority
+   \<Rightarrow> Deterministic_A.ready_queue \<Rightarrow> ready_queue
+   \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref) \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref)
+   \<Rightarrow> (obj_ref \<Rightarrow> bool) \<Rightarrow> bool"
+  where
+  "ready_queue_relation d p q q' nexts prevs flag \<equiv>
+     list_queue_relation q q' nexts prevs
+     \<and> (\<forall>t. flag t \<longleftrightarrow> t \<in> set q)
+     \<and> (d > maxDomain \<or> p > maxPriority \<longrightarrow> tcbQueueEmpty q')"
```

Comment:
Looks like I missed that one before -- it should be a definition, not an abbreviation, because it is unlikely that the term on the right hand side is going to stay in exactly this form after `simp` and other tactics have run. This change is probably necessary for all architectures and might need a few additional unfoldings.

Reply:
I added this fairly late in the game, just to cut down on repetition that appears between `ready_queues_relation` and `ksReadyQueues_asrt` and to make my intentions a bit clearer in the definition I give to `ksReadyQueues_asrt`. I can certainly make this a definition, but seeing that its only appearances are inside other definitions, I don't think we get any unfavourable simplification happening to these bits

Reply:
We could make it an input-only abbreviation -- as a pure abbreviation it doesn't really make sense, it can only work when it is untouched. Actually just unfolding the outer definition could already make the term inapplicable for contraction. As an input abbreviation you'd still get the name to use in other definitions, but would not get any new unfoldings.